### ISSUE TITLE
Lv2 extension plan

### ISSUE THREAD

**Issue by JohannesLorenz** [LMMS/lmms#4954]
** This is WIP **

### Background

Lv2 has many extensions, and some of them contain features. Extensions are things the plugin provides for a host, and features are things the host provides for a plugin. Note: Even if a host does not care about a specific extension, the plugin can require a feature from that extension, which means the plugin won't run without it.

### Purpose

This plan helps to identify
* which extensions we should implement
* in what order

### Table

| Extension | Things to use   | Features to implement | Lilv Support | Depends on   |
|---|---|---|---|---|
|Atom   | ?   | No features exist  | All?  | -   |
|Buf Size    |   |   |   |   |
|Data Access   |None. Discouraged.   |-   |  |   |
|Dynamic Manifest   |   |   |   |   |
|Instance Access   |   |   |   |   |
|Log   |   |   |   |   |
|Lv2   |   |   |   |   |
|Midi   |   |   |   |   |
|Morph   |   |   |   |   |
|Options   |   |   |   |   |
|Parameters   |   |   |   |   |
|Patch   |   |   |   |   |
|Port Groups   |   |   |   |   |
|Port Properties   |   |   |   |   |
|Patch   |   |   |   |   |
|Patch   |   |   |   |   |
|Patch   |   |   |   |   |
|Patch   |   |   |   |   |
|Patch   |   |   |   |   |

**Comment by Spekular** [comment]
@JohannesLorenz URID should be marked as done now, shouldn't it?

**Comment by JohannesLorenz** [comment]
@Spekular Yes, I forgot that. Thanks, fixed now.

**Comment by zonkmachine** [comment]
@JohannesLorenz Did https://github.com/LMMS/lmms/pull/5691 implement all ot the Atom extension?

I've played around with LV2 now and it's brilliant. Thank you so much for working on this!

I installed a whole bunch of plugins via the ubuntu repo and of the ones who wouldn't load into lmms on start the most common message was:
`required feature not supported : http://lv2plug.in/ns/ext/worker#schedule`

Next most are: 
`too many audio output channels ...`

and the 'zyn' stuff calling for:
`required feature not supported : http://lv2plug.in/ns/ext/options#options`

**Comment by JohannesLorenz** [comment]
Thanks @zonkmachine . I updated Atom, we only support MIDI atoms now (all other are discarded). This whole plan may be a bit inexact sometimes.

My next plans are indeed options and buf-size. If worker is easy, we can add that, too. Too many audio output channels: I don't know how this could be fixed best. If you have more than 2, which ones shall be sent to LMMS and which ones shall be ignored?

Btw: Which ubuntu lv2 plugins did you install? Can you please add it on [the Lv2 wiki page](https://github.com/LMMS/lmms/wiki/Lv2)?

**Comment by mxmilkiib** [comment]
I've added a wiki [note](https://github.com/lv2/lv2/wiki#daw--groovebox).

**Comment by enp2s0** [comment]
@JohannesLorenz for the too many outputs case I would route 1 -> left and 2 -> right, ignoring the rest.

**Comment by JohannesLorenz** [comment]
> for the too many outputs case I would route 1 -> left and 2 -> right, ignoring the rest.

I don't think it's a good solution... For many plugins, the use might not make sense or even be unclear if only 2 of >2 audio outputs are connected. Imagine a "2 -> 4 split" plugin that takes a stereo signal and sends out the left signal to the first stereo output and the right signal to the second stereo output. If you only offer one stereo output for this, this plugin will always ignore the right signal, looking like a bug.

**Comment by JohannesLorenz** [comment]
I'll code "worker" next.

**Comment by zonkmachine** [comment]
I marked **Buf Size** done in the list above.

**Comment by JohannesLorenz** [comment]
I've been checking for lv2-plugins which have features we don't support (this can be seen by starting lmms with `LMMS_LV2_DEBUG=1` environment variable: `required feature not supported`). I could not find any except `state#loadDefaultState` and `worker#schedule` (the latter being work in progress). If anyone sees another one, it would be nice if you would write it down here.

**Comment by zonkmachine** [comment]
I haven't been able to find any but some plugins show up with a `can not be loaded:` and no message specified.

Example:
```
Lv2 plugin "Matrix Mixer 8x8" (URI: http://gareus.org/oss/lv2/matrixmixer#i8o8 ) can not be loaded:
Lv2 plugin "MIDI Key-Range Filter" (URI: http://gareus.org/oss/lv2/midifilter#keyrange ) can not be loaded:
Lv2 plugin "MIDI Key-Range Filter" (URI: http://gareus.org/oss/lv2/midifilter#keyrange ) can not be loaded:
Lv2 plugin "MIDI Key-Range Filter" (URI: http://gareus.org/oss/lv2/midifilter#keyrange ) can not be loaded:
```

**Comment by JohannesLorenz** [comment]
> I haven't been able to find any but some plugins show up with a `can not be loaded:` and no further message specified.
> 
> Example:
> 
> ```
> Lv2 plugin "Matrix Mixer 8x8" (URI: http://gareus.org/oss/lv2/matrixmixer#i8o8 ) can not be loaded:
> Lv2 plugin "MIDI Key-Range Filter" (URI: http://gareus.org/oss/lv2/midifilter#keyrange ) can not be loaded:
> Lv2 plugin "MIDI Key-Range Filter" (URI: http://gareus.org/oss/lv2/midifilter#keyrange ) can not be loaded:
> Lv2 plugin "MIDI Key-Range Filter" (URI: http://gareus.org/oss/lv2/midifilter#keyrange ) can not be loaded:
> ```

Actually, it only prints "can not be loaded" if it has at least one reason, so it's really strange you get these. At least, I get messages for them:

```
Lv2 plugin "Matrix Mixer 8x8" (URI: http://gareus.org/oss/lv2/matrixmixer#i8o8 ) can not be loaded:
  -  too many audio input channels : 8
  -  too many audio output channels : 8
Lv2 plugin "MIDI Key-Range Filter" (URI: http://gareus.org/oss/lv2/midifilter#keyrange ) can not be loaded:
  -  no audio output channel
```

**Comment by zonkmachine** [comment]
> Actually, it only prints "can not be loaded" if it has at least one reason,

Yes. I had made some changes to poke a couple of plugins a bit. Reverted that and now I get those messages too. Sorry...

**Comment by zonkmachine** [comment]
> I've been checking for lv2-plugins which have features we don't support

Oh, yes! I'm now running "Super Whammy" from MOD Devices with automation on the knobs. Bliss! All that was needed was to set the buffer to a multiple of 2 and to insert 	**m_supportedFeatureURIs.insert(LV2_BUF_SIZE__powerOf2BlockLength);** in Lv2Manager.cpp . Totally worth it!

**Comment by mxmilkiib** [comment]
I guess #526 and #5795 would come under lv2core?

**Comment by JohannesLorenz** [comment]
> I guess #526 and #5795 (CV ports) would come under lv2core?

#5795: Good point, added.
#526: This looks like a typo? It's not Lv2 related.

**Comment by mxmilkiib** [comment]
Whoops yes, I think I was meaning #5330

**Comment by JohannesLorenz** [comment]
#5330 is interesting, but how come related to Lv2 core? Is there an Lv2 plugin existing with that functionality?

**Comment by mxmilkiib** [comment]
More that it's a basic use of CV interacting with/bridging other track elements, generating note CV from audio (and MIDI) and issuing that to modulate things.

https://github.com/moddevices/mod-cv-plugins/tree/master/source/mod-audio-to-cv

https://github.com/moddevices/mod-cv-plugins/tree/master/source/mod-midi-to-cv-mono

https://github.com/moddevices/mod-cv-plugins/tree/master/source

**Comment by JohannesLorenz** [comment]
Most non-extraordinary extensions seem to be done/in progress.

As I think we should target **UI** ASAP, this should be the next target. Only **state** and **preset** seem to be useful to do before it (not sure yet).

Other interesting stuff seem to be **CV** ports. Other things seem extraordinary to me.

**Comment by stephematician** [comment]
@JohannesLorenz - any progress on the `worker` extension? This is still appearing for me when loading plugins.

[EDIT]Never mind; I found that this has been implemented on master (see PR #6484), but I'm using the last alpha release, and not master or nightly - perhaps I should switch.[/EDIT]

**Comment by JohannesLorenz** [comment]
@stephematician Correct - worker is implemented on master and available in nightlies.
